### PR TITLE
Initial k8s setup

### LINKED ISSUE #41: Initial k8s setup
_wip_

### PR DESCRIPTION
**Pull request by MaxSchaefer**
This pull-request introduces our k8s setup on GKE.
This setup is far away from production ready, but gives us a starting point to build on top.

Directories:
- [/kubernetes/iac](https://github.com/helpwave/services/tree/41-initial-k8s-setup/kubernetes/iac)
	Contains IaC for our GKE cluster on GCP and deploys common/global Helm charts like [Dapr](https://github.com/dapr/helm-charts) and [ingress-nginx](https://github.com/kubernetes/ingress-nginx)
- [/kubernetes/services](https://github.com/helpwave/services/tree/41-initial-k8s-setup/kubernetes/service)
	Contains the Helm chart for our services. Autogenerated with `helm create` and modified for Dapr.
- [/kubernetes/emergency-room-svc](https://github.com/helpwave/services/tree/41-initial-k8s-setup/kubernetes/emergency-room-svc)
	Consumes the services Helm chart and configures that for user-svc. Deploys a [postgresql](https://github.com/bitnami/charts/tree/main/bitnami/postgresql) alongside.
- [/kubernetes/user-svc](https://github.com/helpwave/services/tree/41-initial-k8s-setup/kubernetes/user-svc)
	Consumes the services Helm chart and configures that for emergency-service-svc. Deploys a [postgresql](https://github.com/bitnami/charts/tree/main/bitnami/postgresql) alongside.
- [/kubernetes/nginx-dapr](https://github.com/helpwave/services/tree/41-initial-k8s-setup/kubernetes/nginx-dapr)
	Contains an ingress to map traffic for staging to the Dapr sidecar of nginx. Gets consumed by our IaC.

Moving parts:
- [Terraform](https://github.com/hashicorp/terraform)
	We define our infrastructure on gcp as Iac (Infrastructure-as-Code). Terraform is our tool of choice.
- [helm](https://github.com/helm/helm)
	Templating-Engine and Packet-Manager Kubernetes
- [helmfile](https://github.com/roboll/helmfile)
	Manage multiple Helm charts

Upcoming to-dos which are not part of this pr:
- Move away from helmfile. Migrate fully to Terraform.
- Right now the Helm charts for services are basically the autogenerated from helm with some additions for Dapr and 	migrations. We should clean up these.
- Move away from one PostgreSQL-Server per service.
- Secrets. By whatever method.
- Move IaC to own Repository. Like `helpwave/(iac)`.
- CD to k8s

---

Closes #41

### CODE REVIEW

File: kubernetes/user-svc/helmfile.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  - name: user-postgres
+    namespace: user
+    chart: bitnami/postgresql
+    wait: true
+    set:
+      - name: auth.enablePostgresUser
+        value: true
+      - name: auth.postgresPassword
+        value: helpwave
```
Comment (FoseFx) [inline kubernetes/user-svc/helmfile.yaml:17]: please add "TODO" or "FIXME" comments everywhere we use hardcoded secrets, so it's clear this is not the desired state of the code base

File: kubernetes/nginx-dapr/templates/_helpers.tpl [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+{{/*
+Create a default fully qualified app name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+If release name contains chart name it will be used as a full name.
+*/}}
+{{- define "nginx-dapr.fullname" -}}
+{{- if .Values.fullnameOverride }}
+{{- .Values.fullnameOverride | trunc 63 | trimSuffix "-" }}
+{{- else }}
```
Comment (FoseFx) [inline kubernetes/nginx-dapr/templates/_helpers.tpl:16]: indentation of ifs would be awesome (don't know how w/o breaking yaml though)

File: kubernetes/service/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      serviceAccountName: {{ include "service.serviceAccountName" . }}
+      securityContext:
+        {{- toYaml .Values.podSecurityContext | nindent 8 }}
+      initContainers:
+        - name: migrate
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+          command: ['ash', '-c', 'wget -qO migrate.tgz https://github.com/golang-migrate/migrate/releases/download/v4.15.2/migrate.linux-amd64.tar.gz && tar xzf migrate.tgz && POSTGRESQL_URL="postgres://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST:$POSTGRES_PORT/$POSTGRES_DB?sslmode=disable" && ./migrate -database $POSTGRESQL_URL -path migrations up']
```
Comment (FoseFx) [inline kubernetes/service/templates/deployment.yaml:33]: `ash` = `bash`?

File: kubernetes/service/values.yaml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+podAnnotations: {}
+
+podSecurityContext: {}
+  # fsGroup: 2000
+
+securityContext: {}
+  # capabilities:
+  #   drop:
+  #   - ALL
```
Comment (FoseFx) [inline kubernetes/service/values.yaml:47]: please clean this up

**Review by FoseFx — COMMENTED** [review]
didn't check everything tbh ^^

File: kubernetes/nginx-dapr/values.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Default values for nginx-dapr.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+nginx:
+  ingressDaprSidecar: nginx-ingress-dapr
+
+api:
+  staging:
+    hostname: staging.api.helpwave.de
```
[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply kubernetes/nginx-dapr/values.yaml:10]: Nope, works fine.
https://staging.api.helpwave.de/v1.0/invoke/emergency-room.emergency-room/method/get-emergency-rooms

File: kubernetes/service/templates/deployment.yaml (hunk 3 above)
Comment (MaxSchaefer) [reply kubernetes/service/templates/deployment.yaml:33]: `bash` is not available in `alpine`. We can choose between `ash` and `sh`.

File: kubernetes/iac/gke.tf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  initial_node_count = 2
+  lifecycle {
+    ignore_changes = [
+      initial_node_count
+    ]
+  }
+
+  autoscaling {
+    location_policy = "ANY"
+    min_node_count = 1
```
[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply kubernetes/iac/gke.tf:22]: Nope, but if it did, we would have to scale all nodes according to HTTP ingress requests and not according to cpu utilization. We should develop a slack bot to turn on/off our non-prod clusters. I will write a feature for it, is on my todos.

File: kubernetes/nginx-dapr/values.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply kubernetes/nginx-dapr/values.yaml:6]: Yes, you're right. What would you recommend? I think we want to continue using Cloudflare as a proxy.

File: kubernetes/nginx-dapr/templates/_helpers.tpl (hunk 2 above)
Comment (MaxSchaefer) [reply kubernetes/nginx-dapr/templates/_helpers.tpl:16]: As I said before, most of it was auto-generated by Helm. `helm lint` does not complain either.

File: kubernetes/nginx-dapr/values.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (FoseFx) [reply kubernetes/nginx-dapr/values.yaml:6]: flatten it to `staging-api.helpwave.de` would be the easiest solution I assume

[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply kubernetes/nginx-dapr/values.yaml:6]: @use-to can you do this?
`api.helpwave.de            A    xxx.xxx.xxx.xxx`
`staging-api.helpwave.de    A    xxx.xxx.xxx.xxx`

[the remark this answers is not in the document]
Comment (use-to) [reply kubernetes/nginx-dapr/values.yaml:6]: We're currently not using the cf-proxy. I'll change that...

[the remark this answers is not in the document]
Comment (use-to) [reply kubernetes/nginx-dapr/values.yaml:6]: **Update**: Now, we got `api.helpwave.de`, `*.api.helpwave.de` and `staging-api.helpwave.de` pointing on the same IPv4! CF-Proxy is enabled, too.

[the remark this answers is not in the document]
Comment (use-to) [reply kubernetes/nginx-dapr/values.yaml:6]: https://staging-api.helpwave.de/ is responding...

[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply kubernetes/nginx-dapr/values.yaml:6]: Nice, thanks!
`curl -d '{}'  https://staging-api.helpwave.de/v1.0/invoke/emergency-room.emergency-room/method/get-emergency-rooms`

**Review by FoseFx — APPROVED** [review]

**Pull request merged**